File: db/src/main/java/com/readrops/db/entities/account/AccountConfig.java
package com.readrops.db.entities.account;

public class AccountConfig {

    public static final AccountConfig LOCAL = new AccountConfigBuilder()
            .setFeedUrlEditable(true)
            .setFolderCreation(true)
            .setNoFolderCase(false)
            .setUseSeparateState(false)
            .build();

    public static final AccountConfig NEXTCLOUD_NEWS = new AccountConfigBuilder()
            .setFeedUrlEditable(false)
            .setFolderCreation(true)
            .setNoFolderCase(false)
            .setUseSeparateState(false)
            .build();

    public static final AccountConfig FRESHRSS = new AccountConfigBuilder()
            .setFeedUrlEditable(false)
            .setFolderCreation(false)
            .setNoFolderCase(true)
            .setUseSeparateState(true)
            .build();

    private final boolean feedUrlEditable;

    private final boolean folderCreation;

    private final boolean noFolderCase;

    /*
    Let knows if it uses ItemState table to synchronize state
     */
    private final boolean useSeparateState;

    public boolean isFeedUrlEditable() {
        return feedUrlEditable;
    }

    public boolean isFolderCreation() {
        return folderCreation;
    }

    public boolean isNoFolderCase() {
        return noFolderCase;
    }

    public boolean useSeparateState() {
        return useSeparateState;
    }

    public AccountConfig(AccountConfigBuilder builder) {
        this.feedUrlEditable = builder.feedUrlEditable;
        this.folderCreation = builder.folderCreation;
        this.noFolderCase = builder.noFolderCase;
        this.useSeparateState = builder.useSeparateState;
    }

    public static class AccountConfigBuilder {
        private boolean feedUrlEditable;
        private boolean folderCreation;
        private boolean noFolderCase;
        private boolean useSeparateState;

        public AccountConfigBuilder setFeedUrlEditable(boolean feedUrlEditable) {
            this.feedUrlEditable = feedUrlEditable;
            return this;
        }

        public AccountConfigBuilder setFolderCreation(boolean folderCreation) {
            this.folderCreation = folderCreation;
            return this;
        }

        public AccountConfigBuilder setNoFolderCase(boolean noFolderCase) {
            this.noFolderCase = noFolderCase;
            return this;
        }

        public AccountConfigBuilder setUseSeparateState(boolean useSeparateState) {
            this.useSeparateState = useSeparateState;
            return this;
        }

        public AccountConfig build() {
            return new AccountConfig(this);
        }
    }
}

File: db/src/main/java/com/readrops/db/entities/account/AccountConfig.kt
package com.readrops.db.entities.account

data class AccountConfig(
    val isFeedUrlReadOnly: Boolean, // Enable or disable feed url modification in Feed Tab
    val addNoFolder: Boolean, // Add a "No folder" option when modifying a feed's folder
    val useSeparateState: Boolean, // Let know if it uses ItemState table to synchronize read/star state
    val canCreateFolder: Boolean, // Enable or disable folder creation in Feed Tab
    val canCreateFeed: Boolean = true,
    val canUpdateFolder: Boolean = true,
    val canUpdateFeed: Boolean = true,
    val canDeleteFeed: Boolean = true,
    val canDeleteFolder: Boolean = true,
    val canMarkAllItemsAsRead: Boolean = true
) {

    companion object {
        val LOCAL = AccountConfig(
            isFeedUrlReadOnly = false,
            canCreateFolder = true,
            addNoFolder = true,
            useSeparateState = false,
        )

        val NEXTCLOUD_NEWS = AccountConfig(
            isFeedUrlReadOnly = true,
            canCreateFolder = true,
            addNoFolder = true,
            useSeparateState = false,
        )

        val GREADER = AccountConfig(
            isFeedUrlReadOnly = true,
            canCreateFolder = false,
            addNoFolder = false,
            useSeparateState = true,
        )

        val FEVER = AccountConfig(
            isFeedUrlReadOnly = false,
            addNoFolder = true,
            useSeparateState = true,
            canCreateFolder = false,
            canCreateFeed = false,
            canUpdateFolder = false,
            canUpdateFeed = false,
            canDeleteFeed = false,
            canDeleteFolder = false,
            canMarkAllItemsAsRead = false
        )
    }
}